pkg/statesolver/context_locker.go

```diff
@@ -0,0 +1,56 @@
+package statesolver
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/NiR-/zbuild/pkg/builddef"
+)
+
+var (
+	imageGit = "docker.io/akerouanton/zbuild-git:v0.1"
+)
+
+func LockContext(ctx context.Context, solver StateSolver, c *builddef.Context) (*builddef.Context, error) {
+	if c == nil {
+		return nil, nil
+	}
+
+	// Only git contexts can be locked to a precise version.
+	if !c.IsGitContext() {
+		return c, nil
+	}
+
+	repoURI := normalizeRepoURI(c)
+	sourceRef := sourceRefOrHead(c)
+
+	cmd := []string{
+		fmt.Sprintf("git clone %s /tmp/repo 1>/dev/null 2>&1", repoURI),
+		"cd /tmp/repo",
+		fmt.Sprintf("git rev-parse -q --verify '%s'", sourceRef)}
```

Comment:
isn't it possible to only clone the repo at sourceRef ? (instead of a full clone which will also fetch all the commits) ?

cf https://stackoverflow.com/questions/31278902/how-to-shallow-clone-a-specific-commit-with-depth-1

Reply:
There're no easy way to solve this given that `sourceRef` could be either a branch name, a tag or a sha1. There're other parts to optimize too (e.g. to not clone the same repo multiple times). As such, I'm not sure how all of this will evolve in the next weeks so I'd prefer to keep it simple enough to be easy to refactor :wink: